### PR TITLE
DBZ-971 UnwrapFromMongoDbEnvelope operation detection

### PR DESCRIPTION
**Pull request by renatomefi**
Please check the context at https://issues.jboss.org/browse/DBZ-971

### CODE REVIEW

**Comment by renatomefi** [comment]
@gunnarmorling thanks for the review!

> can you give some background why you'd like to make the distinction between inserts and updates? I.e. what's the rationale behind the change?

Is this context enough or you want more details? https://issues.jboss.org/browse/DBZ-971

> In terms of a possible solution, I'm not a big fan of altering the schema name. How about expressing the operation via a Kafka message header instead? It's transparent to the actual value and its schema that way, while still allowing consumers to react to it as needed.

That sounds good, can you point me to how I can achieve that? Will you be willing the merge it in case I get the feature right?

**Comment by gunnarmorling** [comment]
> Is this context enough or you want more details? https://issues.jboss.org/browse/DBZ-971

I was checking it, but all I found was this:

> I want to be able to tell whether the message is an update or insert thus easing logic within the Sink Connector

So I'm curious abou the "why" of this; how does it exactly ease logic for sink connectors?

> That sounds good, can you point me to how I can achieve that?

Check out the overload of `ConnectRecord.newRecord()` that takes an `Iterable<Header>` as last parameter. The Debezium connectors currently don't add any headers themselves, but we may do it in the future. So it'd be a good idea to propagate any existing headers to the outbound record and add the new one for the operation type.

> Will you be willing the merge it in case I get the feature right?

Yes, if you can explain a bit more why exactly you need it :)

Thanks a lot!

**Comment by gunnarmorling** [comment]
> The rationale is about optimizing the Sink in a way it doesn't have to look for an item before knowing whether it should insert or update, this can potentially save a great number of queries in the Sink side.

Got it. Out of curiosity, what store on the sink side are you using? Many DBs nowawdays have some support for "UPSERT" semantics, be it via SQL extensions or SQL's own `MERGE` statement. But propagating the original op as a header property should be non-disruptive, so +1 for doing it.

> I'm wondering if then I should look for an implementation which works for the whole connector itself and allows the headers to be in there or if I can still do it only for the unwrapper?

Let's just do it for the unwrapping SMT for the time being. We can expand in actual connectors as needed down the road. Just make sure to propagate any existing headers, so there won't be bad surprises if we actually do add headers later on and forget about adjusting that SMT. A nice thing to do would be to adjust the other unwrapping SMT (for relational connectors) equivalently.

**Comment by renatomefi** [comment]
> Got it. Out of curiosity, what store on the sink side are you using? Many DBs nowawdays have some support for "UPSERT" semantics, be it via SQL extensions or SQL's own MERGE statement. But propagating the original op as a header property should be non-disruptive, so +1 for doing it.

In our case it's because we have multiple consumers for this topic (DDD based micro services) thus we want to trigger domain events in the fastest way as possible, and depending on the operation we have already all the data necessary to it! It's not really an storage issue per say!

> Let's just do it for the unwrapping SMT for the time being. We can expand in actual connectors as needed down the road. Just make sure to propagate any existing headers, so there won't be bad surprises if we actually do add headers later on and forget about adjusting that SMT. A nice thing to do would be to adjust the other unwrapping SMT (for relational connectors) equivalently.

Great idea, I've done a regression test on the Mongo one and added the feature, I'll now proceed to add the same for the other unwrap

**Comment by renatomefi** [comment]
@gunnarmorling It's ready, looking forward to your review!

File: debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java [hunk 1]
```diff
@@ -164,13 +180,31 @@ public void configure(final Map<String, ?> configs) {
...
         }
+
+        if (addDebeziumOperationHeader) {
+            try {
+                operation = Envelope.Operation.forCode(
+                        ((Struct) record.value()).get("op").toString()
+                );
+                record.headers().addString(DEBEZIUM_OPERATION_HEADER_KEY, operation.code());
+            }
+            catch (NullPointerException exception) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java]: Catching the NPE seems a bit fishy. I'd rather suggest to check for presence of the "op" field upfront.

File: debezium-core/src/test/java/io/debezium/transforms/UnwrapFromEnvelopeTest.java [hunk 2]
```diff
@@ -218,4 +239,23 @@ public void testIgnoreUnknownRecord() {
             assertThat(transform.apply(unnamedSchemaRecord)).isEqualTo(unnamedSchemaRecord);
         }
     }
+
+    @Test
+    public void testUnwrapPropagatesRecordHeaders() {
```
Comment (gunnarmorling) [inline debezium-core/src/test/java/io/debezium/transforms/UnwrapFromEnvelopeTest.java]: Can you add the `@FixFor` marker annotation?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java [hunk 3]
```diff
@@ -38,6 +38,8 @@
  */
 public class UnwrapFromMongoDbEnvelope<R extends ConnectRecord<R>> implements Transformation<R> {
 
+    final String DEBEZIUM_OPERATION_HEADER_KEY = "application/debezium-operation";
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: So I'm not sure whether there are any established conventions for naming header keys.

But if not, I'd suggest to go with `__debezium-operation`. The reason being that we use the `__debezium` prefix already in a few other places for injecting custom information, e.g. for the schema parameters conveying column source type info (`__debezium.source.column.type`) and the name of the heartbeat topic (`__debezium-heartbeat`). It'll still not be 100% consistent, but I think it's as good as it gets. @jpechane, any thoughts from your side?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java [hunk 4]
```diff
@@ -110,13 +112,23 @@ public static ArrayEncoding parse(String value, String defaultValue) {
             .withDescription("Delimiter to concat between field names from the input record when generating field names for the"
                     + "output record.");
 
+    private static final Field OPERATION_HEADER = Field.create("operation.header")
+            .withDisplayName("Adds the debezium operation into the message header")
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: "Adds a message header representing the applied operation"

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java [hunk 5]
```diff
@@ -110,13 +112,23 @@ public static ArrayEncoding parse(String value, String defaultValue) {
...
+                    "Its key is 'application/debezium-operation'");
+
     private final ExtractField<R> afterExtractor = new ExtractField.Value<R>();
     private final ExtractField<R> patchExtractor = new ExtractField.Value<R>();
     private final ExtractField<R> keyExtractor = new ExtractField.Key<R>();
 
     private MongoDataConverter converter;
     private final Flatten<R> recordFlattener = new Flatten.Value<R>();
 
+    private boolean addDebeziumOperationHeader;
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: I'd name it just `addOperationHeader`.

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java [hunk 6]
```diff
@@ -185,9 +200,19 @@ public void shouldGenerateRecordForUpdateEvent() throws InterruptedException {
...
         SourceRecord record = produced.get(0);
 
+        final Map<String, String> props = new HashMap<>();
+        props.put(OPERATION_HEADER, "true");
+        transformation.configure(props);
+
         // when
         SourceRecord transformed = transformation.apply(record);
 
+        // then assert operation header is insert
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java]:
```suggestion
        // then assert operation header is update
```

File: debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java [hunk 7]
```diff
@@ -261,9 +286,19 @@ public void shouldGenerateRecordForDeleteEvent() throws InterruptedException {
...
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/test/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelopeTest.java]:
```suggestion
        // then assert operation header is delete
```

File: debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java [hunk 8]
```diff
@@ -39,6 +41,8 @@
  */
 public class UnwrapFromEnvelope<R extends ConnectRecord<R>> implements Transformation<R> {
 
+    final String DEBEZIUM_OPERATION_HEADER_KEY = "application/debezium-operation";
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java]: Same comment as above.

File: debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java [hunk 9]
```diff
@@ -115,9 +119,19 @@ public static DeleteHandling parse(String value, String defaultValue) {
                     + "drop - records are removed,"
                     + "rewrite - __deleted field is added to records.");
 
+    private static final Field OPERATION_HEADER = Field.create("operation.header")
+            .withDisplayName("Adds the debezium operation into the message header")
+            .withType(ConfigDef.Type.BOOLEAN)
+            .withWidth(ConfigDef.Width.SHORT)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withDefault(false)
+            .withDescription("Adds the debezium operation as in {@link FieldName#OPERATION operation} to the message header." +
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java]:
```suggestion
            .withDescription("Adds the applied operation as in {@link FieldName#OPERATION operation} to the message header." +
```

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java (hunk 3 above)
Comment (renatomefi) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: @gunnarmorling a ping! :)

Comment (renatomefi) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/UnwrapFromMongoDbEnvelope.java]: I think I went a bit off, I talked to a good Kafka person and I was told the HTTP is not really a reference there.
So I changed it to your suggestion :+1:

**Comment by renatomefi** [comment]
Closed in favor of #682

**Pull request closed without merging**